File: backend/src/infrastructure/crypto/crypto.service.ts
// AES-256 Encryption Service
// Section 5 of Master Prompt - MANDATORY SECURITY RULE

import { Injectable, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as crypto from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 16;
const TAG_LENGTH = 16;
const ENC_PREFIX = 'ENC:';

@Injectable()
export class CryptoService implements OnModuleInit {
  private encryptionKey: Buffer;

  constructor(private readonly configService: ConfigService) {}

  onModuleInit() {
    const key = this.configService.get<string>('ENCRYPTION_KEY');
    if (!key) {
      throw new Error('ENCRYPTION_KEY environment variable is required');
    }
    if (key.length < 32) {
      throw new Error('ENCRYPTION_KEY must be at least 32 characters (256 bits)');
    }
    this.encryptionKey = Buffer.from(key.slice(0, 32), 'utf-8');
  }

  /**
   * Encrypts plaintext using AES-256-GCM
   * Returns: IV (16 bytes) + AuthTag (16 bytes) + Ciphertext, base64 encoded
   */
  encrypt(plaintext: string): string {
    return CryptoService.encryptWithKey(plaintext, this.encryptionKey);
  }

  /**
   * Decrypts ciphertext encrypted with AES-256-GCM
   * Input: IV (16 bytes) + AuthTag (16 bytes) + Ciphertext, base64 encoded
   */
  decrypt(encryptedData: string): string {
    return CryptoService.decryptWithKey(encryptedData, this.encryptionKey);
  }

  /**
   * Decrypts an env value if it has the ENC: prefix, otherwise returns as-is
   */
  decryptEnv(value: string | undefined): string {
    if (!value) return '';
    if (value.startsWith(ENC_PREFIX)) {
      return this.decrypt(value.slice(ENC_PREFIX.length));
    }
    return value;
  }

  /**
   * Hashes data using SHA-256
   */
  hash(data: string): string {
    return crypto.createHash('sha256').update(data).digest('hex');
  }

  // ── Static methods for standalone usage (outside NestJS DI) ──

  /**
   * Encrypts plaintext using a raw key buffer
   */
  static encryptWithKey(plaintext: string, keyBuffer: Buffer): string {
    const iv = crypto.randomBytes(IV_LENGTH);
    const cipher = crypto.createCipheriv(ALGORITHM, keyBuffer, iv);

    let encrypted = cipher.update(plaintext, 'utf8');
    encrypted = Buffer.concat([encrypted, cipher.final()]);

    const authTag = cipher.getAuthTag();
    const combined = Buffer.concat([iv, authTag, encrypted]);
    return combined.toString('base64');
  }

  /**
   * Decrypts ciphertext using a raw key buffer
   */
  static decryptWithKey(encryptedData: string, keyBuffer: Buffer): string {
    const combined = Buffer.from(encryptedData, 'base64');

    const iv = combined.subarray(0, IV_LENGTH);
    const authTag = combined.subarray(IV_LENGTH, IV_LENGTH + TAG_LENGTH);
    const ciphertext = combined.subarray(IV_LENGTH + TAG_LENGTH);

    const decipher = crypto.createDecipheriv(ALGORITHM, keyBuffer, iv);
    decipher.setAuthTag(authTag);

    let decrypted = decipher.update(ciphertext);
    decrypted = Buffer.concat([decrypted, decipher.final()]);

    return decrypted.toString('utf8');
  }

  /**
   * Builds a key buffer from a string (for standalone use)
   */
  static buildKey(keyString: string): Buffer {
    if (!keyString || keyString.length < 32) {
      throw new Error('ENCRYPTION_KEY must be at least 32 characters');
    }
    return Buffer.from(keyString.slice(0, 32), 'utf-8');
  }

  /**
   * Static decrypt for env values with ENC: prefix (standalone, no DI needed)
   */
  static decryptEnvValue(value: string | undefined, encryptionKey: string): string {
    if (!value) return '';
    if (value.startsWith(ENC_PREFIX)) {
      const keyBuffer = CryptoService.buildKey(encryptionKey);
      return CryptoService.decryptWithKey(value.slice(ENC_PREFIX.length), keyBuffer);
    }
    return value;
  }

  static readonly ENC_PREFIX = ENC_PREFIX;
}

File: python-service/app/services/crypto_service.py
"""
AES-256-GCM Encryption Service
Mirrors Backend's cryptography implementation
"""
import os
import base64
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


ALGORITHM = 'aes-256-gcm'
IV_LENGTH = 16
TAG_LENGTH = 16
ENC_PREFIX = 'ENC:'


class CryptoService:
    """AES-256-GCM Encryption/Decryption Service"""

    def __init__(self):
        """Initialize with ENCRYPTION_KEY from environment"""
        self.encryption_key = self._load_key()

    def _load_key(self) -> bytes:
        """Load and validate encryption key from environment"""
        key = os.getenv('ENCRYPTION_KEY', '')
        if not key:
            raise ValueError('ENCRYPTION_KEY environment variable is required')
        if len(key) < 32:
            raise ValueError('ENCRYPTION_KEY must be at least 32 characters (256 bits)')
        # Use first 32 characters (256 bits)
        return key[:32].encode('utf-8')

    @staticmethod
    def decrypt_with_key(encrypted_data: str, key_buffer: bytes) -> str:
        """
        Decrypt ciphertext encrypted with AES-256-GCM
        Input: IV (16 bytes) + AuthTag (16 bytes) + Ciphertext, base64 encoded
        """
        try:
            combined = base64.b64decode(encrypted_data)

            iv = combined[:IV_LENGTH]
            auth_tag = combined[IV_LENGTH:IV_LENGTH + TAG_LENGTH]
            ciphertext = combined[IV_LENGTH + TAG_LENGTH:]

            cipher = Cipher(
                algorithms.AES(key_buffer),
                modes.GCM(iv, auth_tag),
                backend=default_backend()
            )
            decryptor = cipher.decryptor()

            decrypted = decryptor.update(ciphertext) + decryptor.finalize()
            return decrypted.decode('utf-8')
        except Exception as e:
            raise ValueError(f'Decryption failed: {str(e)}')

    def decrypt(self, encrypted_data: str) -> str:
        """Decrypt using service's encryption key"""
        return self.decrypt_with_key(encrypted_data, self.encryption_key)

    @staticmethod
    def encrypt_with_key(plaintext: str, key_buffer: bytes) -> str:
        """
        Encrypt plaintext using AES-256-GCM
        Output: IV (16 bytes) + AuthTag (16 bytes) + Ciphertext, base64 encoded
        Mirrors Backend's implementation
        """
        try:
            iv = os.urandom(IV_LENGTH)
            cipher = Cipher(
                algorithms.AES(key_buffer),
                modes.GCM(iv),
                backend=default_backend()
            )
            encryptor = cipher.encryptor()

            encrypted = encryptor.update(plaintext.encode('utf-8')) + encryptor.finalize()
            auth_tag = encryptor.tag

            combined = iv + auth_tag + encrypted
            return base64.b64encode(combined).decode('utf-8')
        except Exception as e:
            raise ValueError(f'Encryption failed: {str(e)}')

    def encrypt(self, plaintext: str) -> str:
        """Encrypt using service's encryption key"""
        return self.encrypt_with_key(plaintext, self.encryption_key)

    @staticmethod
    def decrypt_env_value(value: str | None, encryption_key: str) -> str:
        """
        Decrypt an env value if it has the ENC: prefix, otherwise return as-is
        """
        if not value:
            return ''
        if value.startswith(ENC_PREFIX):
            key_buffer = encryption_key[:32].encode('utf-8')
            return CryptoService.decrypt_with_key(value[len(ENC_PREFIX):], key_buffer)
        return value

    def decrypt_env(self, value: str | None) -> str:
        """Decrypt env value using service's key"""
        return self.decrypt_env_value(value, self.encryption_key.decode('utf-8'))


# Global instance
_crypto_service = None


def get_crypto_service() -> CryptoService:
    """Get or create global crypto service instance"""
    global _crypto_service
    if _crypto_service is None:
        _crypto_service = CryptoService()
    return _crypto_service
